### PR TITLE
Problem: no address discovery after wallet recovery#1396

### PR DESCRIPTION
**Pull request by leejw51crypto**
## why ?
`balance` is the most important data which user see.
after hd wallet restoration, user cannot see any balance after sync,because address are not restored.
for testnet, sync usually takes a lot of time and energy.
this is crucial for general users.
address deterministic is one of cool features why we're using hdwallet.
it's very handy to general users
also in future pr, supporting hw wallet will be added

## what i modified
modified two important issues which is found in prior review.
future hardware-wallet support is also considered
pubkey is peeked by parent extended pubkey + index
(no more getting privatekey)

## why i think i fixed the problem?
- restore wallet
- sync
- i can check latest balance
- also can do the same after the wallet duplications

## modifications after previous PR 

1. remove private keypair calling, just calling get_pubkey
so current discovery logic can be reused for hardware wallet.
only latest index should be store somewhere for hw wallet though, but that's necessary for adress discovery and normal hw wallet use case.

2. not use WalletClient in syncing, only AddressRecovery trait, in syncing, using privatekey is not allowed,
because new transfer address also generates
- multisig session
- multisig roothash -> utxo address
- corresponding pubkey
- above data are related , so it makes sense to reuse just current code-base

3. also get pub_key(index) is being used, hw wallet friendly

### CODE REVIEW

**Comment by leejw51crypto** [comment]
writing unit-test

File: client-cli/src/command.rs [hunk 1]
```diff
@@ -652,3 +668,29 @@ fn print_sync_warning() {
     ask("Warning! Information displayed here may be outdated. To get the latest information, do `client-cli sync --name <wallet name>`");
     println!();
 }
+
+fn get_wallet_client() -> Result<AppWalletClient> {
+    let tendermint_client = WebsocketRpcClient::new(&tendermint_url())?;
+    let _tx_obfuscation = get_tx_query(tendermint_client.clone())?;
+    let _enckey = ask_seckey(None)?;
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply client-cli/src/command.rs:675]: i'll check

File: client-core/src/wallet/syncer.rs [hunk 2]
```diff
@@ -20,6 +20,13 @@ use super::syncer_logic::handle_blocks;
 use crate::service;
 use crate::service::{KeyService, SyncState, Wallet, WalletState, WalletStateMemento};
 use crate::TransactionObfuscation;
+use chain_core::tx::TransactionId;
+use std::collections::HashMap;
+
+pub trait AddressRecovery: Clone + Send + Sync {
+    fn recover_addresses(&mut self, new_address: &str, name: &str, enckey: &SecKey)
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply client-core/src/wallet/syncer.rs:27]: i'll add

File: client-core/src/wallet/syncer.rs [hunk 3]
```diff
@@ -277,14 +305,78 @@ impl<'a, S: SecureStorage, C: Client, D: TxDecryptor, F: FnMut(ProgressReport) -
...
 
+    fn handle_recover_addresses_for_transaction(
+        &mut self,
+        transaction: &Transaction,
+    ) -> Result<()> {
+        let _transaction_id = transaction.id();
+
+        let mut refetch = false;
+
+        let _inputs = transaction.inputs().to_vec();
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply client-core/src/wallet/syncer.rs:316]: i'll check

File: client-rpc/src/rpc/sync_rpc.rs [hunk 4]
```diff
@@ -3,7 +3,7 @@ use super::sync_worker::WorkerShared;
 use crate::server::to_rpc_error;
 use client_common::tendermint::Client;
 use client_common::Storage;
-use client_core::synchronizer::PollingSynchronizer;
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply client-rpc/src/rpc/sync_rpc.rs:6]: this is not used any more,i'll add to commit message

File: client-core/src/service/root_hash_service.rs [hunk 5]
```diff
@@ -22,6 +22,21 @@ where
         Self { storage }
     }
 
+    /// peek new root_hash
+    pub fn peek_new_root_hash(
+        &self,
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply client-core/src/service/root_hash_service.rs:27]: yes, it's not necessary
i'll remove it

**Comment by leejw51crypto** [comment]
> looks cleaner than the previous PR.
> I guess this is to be merged after #1496 ?
> 
> it'll also be good to summarise the solution/required changes in the relevant part of the commit message -- currently it says "Solution: add address recovery in sync"
> which isn't that useful -- it'll be good to summarise this uses the parent pubkey index;
> also why polling sync was removed?
> 
> and
> why is clone required on all those different traits? (or maybe it isn't required -- in that case, just change back those?)

i'll add more detail to commit message

File: client-cli/src/command.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply client-cli/src/command.rs]: cloning is for multi-thread sync, but changed back WalletClient

**Comment by leejw51crypto** [comment]
modified for above feedback,
now writing unit test

File: client-core/src/wallet/syncer.rs [hunk 6]
```diff
@@ -277,14 +307,74 @@ impl<'a, S: SecureStorage, C: Client, D: TxDecryptor, F: FnMut(ProgressReport) -
...
+        transaction: &Transaction,
+    ) -> Result<()> {
+        let mut refetch = false;
+
+        let outputs = transaction.outputs().to_vec();
+
+        for (_i, output) in outputs.iter().enumerate() {
+            let newaddress = output.address.to_string();
+            let tmp_refetch = self.env.recover_address.recover_addresses(
+                &newaddress,
```
Comment (tomtau) [inline client-core/src/wallet/syncer.rs:321]: `output.address` could be passed directly?

File: client-core/src/wallet/default_wallet_client.rs [hunk 7]
```diff
@@ -110,6 +112,71 @@ where
     }
 }
 
+impl<S, C, T> AddressRecovery for DefaultWalletClient<S, C, T>
+where
+    S: Storage,
+    C: Client,
+    T: WalletTransactionBuilder,
+{
+    // new_address: transfer address in TxOut, it will check whether it belongs with 20 window, then it will create 20 new addresses
```
[the remark this answers is not in the document]
Comment (leejw51) [reply client-core/src/wallet/default_wallet_client.rs:121]: ok,i'll change it

[the remark this answers is not in the document]
Comment (leejw51crypto) [reply client-core/src/wallet/default_wallet_client.rs:121]: ok

File: client-core/src/wallet/syncer.rs (hunk 6 above)
Comment (leejw51crypto) [reply client-core/src/wallet/syncer.rs:321]: yes,i'll change it

**Comment by leejw51crypto** [comment]
done,
rebasing with master

**Pull request merged**